File: ios_native/SmartExpenseManager/SmartExpenseManager/Presentation/Features/Expense/ViewModels/ExpenseViewModel.swift
import Foundation
import SwiftUI

@MainActor
class ExpenseViewModel: ObservableObject {
    @Published var expenses: [Expense] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var selectedCategory: Category?
    @Published var selectedDate: Date = Date()

    private let getExpensesUseCase: GetExpensesUseCase
    private let saveExpenseUseCase: SaveExpenseUseCase
    private let deleteExpenseUseCase: DeleteExpenseUseCase

    init(
        getExpensesUseCase: GetExpensesUseCase,
        saveExpenseUseCase: SaveExpenseUseCase,
        deleteExpenseUseCase: DeleteExpenseUseCase
    ) {
        self.getExpensesUseCase = getExpensesUseCase
        self.saveExpenseUseCase = saveExpenseUseCase
        self.deleteExpenseUseCase = deleteExpenseUseCase
    }

    func loadExpenses() async {
        isLoading = true
        errorMessage = nil

        do {
            expenses = try await getExpensesUseCase.execute()
        } catch {
            errorMessage = "\(AppStrings.failedToLoadExpenses): \(error.localizedDescription)"
        }

        isLoading = false
    }

    func loadExpenses(forDate date: Date) async {
        isLoading = true
        errorMessage = nil

        do {
            expenses = try await getExpensesUseCase.execute(forDate: date)
        } catch {
            errorMessage = "\(AppStrings.failedToLoadExpenses): \(error.localizedDescription)"
        }

        isLoading = false
    }

    func loadExpenses(inCategory category: Category) async {
        isLoading = true
        errorMessage = nil

        do {
            expenses = try await getExpensesUseCase.execute(inCategory: category)
        } catch {
            errorMessage = "\(AppStrings.failedToLoadExpenses): \(error.localizedDescription)"
        }

        isLoading = false
    }

    func saveExpense(
        amount: Double,
        category: Category,
        date: Date,
        note: String?
    ) async {
        isLoading = true
        errorMessage = nil

        let expense = Expense(
            amount: amount,
            category: category,
            date: date,
            note: note
        )

        do {
            try await saveExpenseUseCase.execute(expense)
            await loadExpenses()
        } catch {
            errorMessage = "\(AppStrings.failedToSaveExpense): \(error.localizedDescription)"
        }

        isLoading = false
    }

    func deleteExpense(_ expense: Expense) async {
        isLoading = true
        errorMessage = nil

        do {
            try await deleteExpenseUseCase.execute(expense)
            await loadExpenses()
        } catch {
            errorMessage = "\(AppStrings.failedToDeleteExpense): \(error.localizedDescription)"
        }

        isLoading = false
    }

    var totalAmount: Double {
        expenses.reduce(0) { $0 + $1.amount }
    }

    var expensesByCategory: [Category: [Expense]] {
        Dictionary(grouping: expenses) { $0.category }
    }
}

File: android_native/app/src/main/java/com/smartexpense/manager/presentation/viewmodel/ExpenseViewModel.kt
package com.smartexpense.manager.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.smartexpense.manager.domain.model.Category
import com.smartexpense.manager.domain.model.Expense
import com.smartexpense.manager.domain.usecase.DeleteExpense
import com.smartexpense.manager.domain.usecase.GetExpenses
import com.smartexpense.manager.domain.usecase.SaveExpense
import com.smartexpense.manager.presentation.constants.AppStrings
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject

@HiltViewModel
class ExpenseViewModel @Inject constructor(
    private val getExpenses: GetExpenses,
    private val saveExpense: SaveExpense,
    private val deleteExpense: DeleteExpense
) : ViewModel() {

    private val _uiState = MutableStateFlow<ExpenseUiState>(ExpenseUiState.Loading)
    val uiState: StateFlow<ExpenseUiState> = _uiState.asStateFlow()

    init {
        loadExpenses()
    }

    fun loadExpenses() {
        viewModelScope.launch {
            _uiState.value = ExpenseUiState.Loading
            getExpenses().fold(
                onSuccess = { expenses ->
                    val totalAmount = expenses.sumOf { it.amount }
                    _uiState.value = ExpenseUiState.Success(expenses, totalAmount)
                },
                onFailure = { error ->
                    _uiState.value = ExpenseUiState.Error(error.message ?: AppStrings.UNKNOWN_ERROR)
                }
            )
        }
    }

    fun loadExpensesByDate(date: Date) {
        viewModelScope.launch {
            _uiState.value = ExpenseUiState.Loading
            getExpenses.callByDate(date).fold(
                onSuccess = { expenses ->
                    val totalAmount = expenses.sumOf { it.amount }
                    _uiState.value = ExpenseUiState.Success(expenses, totalAmount)
                },
                onFailure = { error ->
                    _uiState.value = ExpenseUiState.Error(error.message ?: AppStrings.UNKNOWN_ERROR)
                }
            )
        }
    }

    fun loadExpensesByCategory(category: Category) {
        viewModelScope.launch {
            _uiState.value = ExpenseUiState.Loading
            getExpenses.callByCategory(category).fold(
                onSuccess = { expenses ->
                    val totalAmount = expenses.sumOf { it.amount }
                    _uiState.value = ExpenseUiState.Success(expenses, totalAmount)
                },
                onFailure = { error ->
                    _uiState.value = ExpenseUiState.Error(error.message ?: AppStrings.UNKNOWN_ERROR)
                }
            )
        }
    }

    fun addExpense(amount: Double, category: Category, date: Date, note: String?) {
        viewModelScope.launch {
            _uiState.value = ExpenseUiState.Saving
            val expense = Expense(
                id = java.util.UUID.randomUUID().toString(),
                amount = amount,
                category = category,
                date = date,
                note = note
            )
            saveExpense(expense).fold(
                onSuccess = {
                    _uiState.value = ExpenseUiState.Saved
                    loadExpenses()
                },
                onFailure = { error ->
                    _uiState.value = ExpenseUiState.Error(error.message ?: AppStrings.FAILED_TO_SAVE_EXPENSE)
                }
            )
        }
    }

    fun deleteExpense(expense: Expense) {
        viewModelScope.launch {
            _uiState.value = ExpenseUiState.Deleting
            deleteExpense(expense).fold(
                onSuccess = {
                    _uiState.value = ExpenseUiState.Deleted
                    loadExpenses()
                },
                onFailure = { error ->
                    _uiState.value = ExpenseUiState.Error(error.message ?: AppStrings.FAILED_TO_DELETE_EXPENSE)
                }
            )
        }
    }
}

sealed class ExpenseUiState {
    object Loading : ExpenseUiState()
    data class Success(val expenses: List<Expense>, val totalAmount: Double) : ExpenseUiState()
    object Saving : ExpenseUiState()
    object Saved : ExpenseUiState()
    object Deleting : ExpenseUiState()
    object Deleted : ExpenseUiState()
    data class Error(val message: String) : ExpenseUiState()
}
